src/formatter.rs

```diff
@@ -40,6 +40,18 @@ pub trait Formatter: Write {
     fn remove_label(&mut self) -> io::Result<()>;
 }
 
+impl dyn Formatter + '_ {
+    pub fn with_label(
+        &mut self,
+        label: &str,
+        write_inner: impl FnOnce(&mut dyn Formatter) -> io::Result<()>,
+    ) -> io::Result<()> {
+        self.add_label(label)?;
+        write_inner(self)?;
+        self.remove_label()
```

Comment:
It needs to call `remove_label()` even if `write_inner()` failed.
`Drop` would help here to ensure `remove_label()` is called (and I have unsent patch to add `impl Drop for ColorFormatter`), but as you said, we can't propagate the io::Error occurred in `drop()`.

Regarding the function signature, I think it's okay to expect the error type is `CommandError` so long as the formatter lives in the cli crate. And the function can be moved out of the object (e.g. `cli_util::with_label(f: &mut F, ..)`) to support concrete `F: Formatter` type if you like.

Reply:
> It needs to call `remove_label()` even if `write_inner()` failed.

I considered doing that but figured it wouldn't be necessary since it would probably fail anyway, but you're right, it's better to still try it. Done.

> Regarding the function signature, I think it's okay to expect the error type is `CommandError` so long as the formatter lives in the cli crate. And the function can be moved out of the object (e.g. `cli_util::with_label(f: &mut F, ..)`) to support concrete `F: Formatter` type if you like.

Yes, that's fair. I prefer to leave it for later since we don't really need it yet. We may want to make the error a generic parameter when we do, so `CommandError` doesn't need to spread to so many places where it can't currently happen.

Reply:
The order of evaluation is deterministic, so this `.and()` should be good.
https://doc.rust-lang.org/reference/expressions.html#evaluation-order-of-operands